### PR TITLE
Tighten output types for `tf.linalg.trace`, `tf.sequence_mask`, and `tf.nn.embedding_lookup`

### PR DESCRIPTION
**Pull request by khatchad**
Tightens the tensor generators that previously left output shape at ⊤ (or emitted a default dtype unconditionally), retiring the precision-limitation list in wala/ML#449. Every op named in that list now has a precise (or maximally-precise) output type:

- **`tf.linalg.trace`**—shape `x.shape[:-2]`.
- **`tf.sequence_mask`**—`dtype` override flows through; shape `lengths.shape + [maxlen]` for constant `maxlen`.
- **`tf.nn.embedding_lookup`**—shape `ids.shape + params.shape[1:]`.
- **`tf.gather_nd`**—shape `indices.shape[:-1] + params.shape[indices.shape[-1]:]` (constant index depth).
- **`tf.linalg.tensordot`**—shape `a.shape[:-n] + b.shape[n:]` for a scalar `axes=n` (list `axes` stays ⊤).
- **`tf.image.extract_patches`**—shape `[batch, out_rows, out_cols, sizes_r * sizes_c * channels]` via the windowed-extent arithmetic over `sizes`/`strides`/`rates`/`padding`.
- **`tf.boolean_mask`**—the masked axis is the runtime `True` count, so it is emitted as a single `Dynamic` dimension with the surrounding dimensions precise (e.g. `[Dynamic, 2]`); this is the maximally-precise static answer.

Pinned with `consume`-sink fixtures; previously-⊤ assertions updated to the precise types. Full suite green.

### CODE REVIEW

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/BooleanMask.java [hunk 1]
```diff
@@ -38,8 +44,48 @@ protected String getInputParameterName() {
...
+      for (List<Dimension<?>> mask : maskShapes) {
+        int maskRank = mask.size();
+        if (axis < 0 || maskRank < 1 || axis + maskRank > tensor.size()) continue;
+        List<Dimension<?>> out = new ArrayList<>(tensor.subList(0, axis));
+        out.add(DynamicDim.INSTANCE);
+        out.addAll(tensor.subList(axis + maskRank, tensor.size()));
+        ret.add(out);
+      }
+    }
+    return ret.isEmpty() ? null : ret;
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/BooleanMask.java]: > `axisPts` can contain multiple constant values (or `None` plus a constant) across contexts, but the current logic overwrites `axis` in a loop and then computes shapes using only the last value. That is unsound and can produce an incorrect (overly precise) shape set when `axis` is ambiguous. Collect all constant axis values (treat `None` as 0) and union the resulting shapes, or return ⊤ when ambiguity cannot be resolved soundly.

Fixed in d3a3a9e1 — the constant `axis` value(s) are now collected into a set (with `None` → 0) and the output is produced per value, so multiple distinct `axis` constants across contexts union rather than the last one winning.

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/BooleanMask.java [hunk 2]
```diff
@@ -1,19 +1,25 @@
...
  * Generator for {@code tf.boolean_mask}. Output dtype is inherited from the {@code tensor} input.
- * Output shape is left at ⊤ for now: the masked dimension's length is the number of {@code True}
- * entries in {@code mask}, a runtime quantity that static analysis cannot recover. The remaining
- * dimensions follow {@code tensor.shape} (offset by {@code axis}), so the result rank is generally
- * {@code tensor.rank - mask.rank + 1}, not always 1-D — but the dependence on the mask's runtime
- * truthiness still leaves the masked dimension undetermined. See wala/ML#449 (Tier 8).
+ * Output shape collapses the masked axes ({@code axis .. axis + mask.rank}) into a single dimension
+ * whose length is the number of {@code True} entries in {@code mask} — a runtime quantity static
+ * analysis cannot recover — keeping the rest of {@code tensor.shape}. That collapsed dimension is
+ * therefore emitted as a {@link DynamicDim}; the rank and the surrounding dimensions are precise.
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/BooleanMask.java]: > The class Javadoc describes the collapsed axis range as `axis .. axis + mask.rank`, but the implementation (and TF semantics) replace exactly `mask.rank` axes starting at `axis`, i.e., `axis .. axis + mask.rank - 1`. Updating the wording avoids an off-by-one misunderstanding for future maintenance.

Clarified in d3a3a9e1 — the Javadoc now describes the collapsed range as the half-open `[axis, axis + mask.rank)` (exactly `mask.rank` axes).

File: com.ibm.wala.cast.python.test/data/tf2_test_sequence_mask_dtype.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+
+def f(a):
+    pass
+
+
+# Counterpart of `tf2_test_sequence_mask.py` that passes the optional `dtype`
+# override (`tf.sequence_mask(..., dtype=tf.int32)`). The output dtype follows
+# the argument (int32 here) instead of the default `bool`. Shape stays at the
+# documented (3, 5) but is not yet inferred precisely.
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.test/data/tf2_test_sequence_mask_dtype.py]: > This comment says the (3, 5) shape "is not yet inferred precisely", but the fixture asserts a precise shape and the PR’s intent is to infer it when `maxlen` is constant. Please update the wording to avoid contradicting the test and the generator behavior.

Fixed in d3a3a9e1 — the fixture comment now says the (3, 5) shape is derived from `lengths` and the constant `maxlen`, rather than "not yet inferred precisely".

**Comment by khatchad** [comment]
> :x: Patch coverage is `73.52941%` with `54 lines` in your changes missing coverage.

For context (the `codecov/patch` and `codecov/project` checks both pass): the missing lines are almost entirely the defensive ⊤-degradation arms in the new generators, not untested logic. Each `getDefaultShapes` returns ⊤ (`null`) when an operand dimension is non-constant, when an argument's points-to set is empty, or when a shape/axis argument can't be resolved to a constant—the sound fallback the lattice conventions require. The `consume`-sink fixtures pin the precise happy-path shapes, so the *partials* Codecov reports are the branch arms that take those `null` fallbacks.

`BooleanMask` and `ExtractPatches` show the most because they have the most such guards (multi-axis mask loops; windowed-extent arithmetic over `VALID`/`SAME` padding). Covering each remaining arm would need a bespoke non-constant-operand fixture per branch for marginal value, since the arms are simple "give up to ⊤" guards. This is the expected coverage shape for shape-derivation generators that are precise on the modeled case and ⊤ elsewhere.
